### PR TITLE
[Resource] Created InMemoryRepository and implemented ResourceInterface

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws \InvalidArgumentException
+     * @throws UnexpectedTypeException
+     */
+    public function add(ResourceInterface $resource)
+    {
+        if (!in_array($this->interface, class_implements($resource))) {
+            throw new UnexpectedTypeException($resource, $this->interface);
+        }
+
+        if (null === $resource->getId()) {
```
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Why do we need to have id in order to add resource to database? If we store resources in memory, we can compare them with `===` or using [spl_object_hash](http://php.net/manual/pl/function.spl-object-hash.php).

File: src/Sylius/Component/Payment/Model/PaymentMethodInterface.php [hunk 2]
```diff
@@ -11,6 +11,7 @@
 
 namespace Sylius\Component\Payment\Model;
 
+use Sylius\Component\Resource\Model\ResourceInterface;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Payment/Model/PaymentMethodInterface.php]: Do we need this?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+     */
+    protected $arrayObject;
+
+    /**
+     * @var string
+     */
+    protected $interface;
+
+    /**
+     * @param string $interface
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: I would add a comment that it should be *FQCN* of the interface.

File: src/Sylius/Component/Payment/Model/PaymentMethodInterface.php (hunk 2 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Payment/Model/PaymentMethodInterface.php]: I think we are missing ``extends`` here. :)

Comment (pjedrzejewski) [inline src/Sylius/Component/Payment/Model/PaymentMethodInterface.php]: Ah no, because it extends PaymentMethodTranslationInterface, I don't like that. But let's fix that in a separate PR I think. :)

File: src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+     */
+    public function getName();
+
+    /**
+     * @return int
+     */
+    public function getRating();
+}
+
+class Repositable implements RepositableInterface
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php]: Classes that are uses for spec should be created in `.../spec/Fixtures/` (see https://github.com/pjedrzejewski/Sylius/blob/introduce-resource-factories/src/Sylius/Component/Inventory/spec/Fixtures/SampleStockable.php).

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 3 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1:

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+     * @throws \InvalidArgumentException
+     * @throws UnexpectedTypeException
+     */
+    public function __construct($interface)
+    {
+        if (null === $interface) {
+            throw new \InvalidArgumentException('Resource\'s interface needs to be stated.');
+        }
+
+        if (!in_array(ResourceInterface::class, class_implements($interface))) {
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1:

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * @param ResourceInterface $resource
+     *
+     * @throws \InvalidArgumentException
+     * @throws UnexpectedTypeException
+     */
+    public function add(ResourceInterface $resource)
+    {
+        if (!in_array($this->interface, class_implements($resource))) {
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: I think we could use ``instanceof`` here, no?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 1 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: If we do not require the id, then we have a problem how ``find`` etc. should work. I think we have a tough problem here...

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+     * {@inheritdoc}
+     *
+     * @throws \UnexpectedValueException
+     */
+    public function findOneBy(array $criteria)
+    {
+        $results = $this->applyCriteria($this->findAll(), $criteria);
+
+        if (1 < $length = count($results)) {
+            throw new \UnexpectedValueException(sprintf('Non unique result. Number of results: %s.', $length));
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1:

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+     *
+     * @return ResourceInterface[]|array
+     */
+    private function applyCriteria(array $resources, array $criteria = array())
+    {
+        foreach ($this->arrayObject as $object) {
+            foreach ($criteria as $criterion => $value) {
+                if ($value !== $this->accessor->getValue($object, $criterion)) {
+                    $key = array_search($object, $resources);
+                    unset($resources[$key]);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Have you tried to achive this by using `array_filter`?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+     */
+    public function findOneBy(array $criteria)
+    {
+        $results = $this->applyCriteria($this->findAll(), $criteria);
+
+        if (1 < $length = count($results)) {
+            throw new \UnexpectedValueException(sprintf('Non unique result. Number of results: %s.', $length));
+        }
+
+        return $results[0];
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: What if there is no results?

File: src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    {
+        $this->getClassName()->shouldReturn(RepositableInterface::class);
+    }
+}
+
+class Void
+{
+}
+
+interface RepositableInterface extends ResourceInterface
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php]: SampleResource?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 1 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: But... since we got autogenerated ids, it won't work with them - so either we must set ids by ourselves earlier or find out another way to handle it. 

I'm for generating ids beforehand, cause we use ids pretty much everywhere and depending on database and flushing in order to get an id is not the best solution.

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 7 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Probably `%d` instead of `%s`, just to be more specific :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+     */
+    public function findAll()
+    {
+        return array_values($this->arrayObject->getArrayCopy());
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function findBy(array $criteria = array(), array $orderBy = null, $limit = null, $offset = null)
```
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Shouldn't `$offset` be `0` by default?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \UnexpectedValueException
+     */
+    public function findOneBy(array $criteria)
+    {
+        $results = $this->applyCriteria($this->findAll(), $criteria);
+
+        if (1 < $length = count($results)) {
```
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: `$length` -> `$count`, `$amount`?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 7 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Anyway, isn't `findOneBy` default behaviour set to return first found element that fulfills requirements?

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+     */
+    public function getClassName()
+    {
+        return $this->interface;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createPaginator(array $criteria = null, array $orderBy = null)
```
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Not sure about `array $variable = null`, why not `array $variable = []`? There shouldn't be performance difference between `isset` and `empty` language constructs.

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 9 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Btw. you can use `reset($results)` to obtain the first result (of course add a check for empty results too).

Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Beware that `reset()` returns `false` if the array is empty :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 6 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: It would be a lot faster :+1:

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 5 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: We can use `instanceof` here too :)

**Comment by michalmarcinkowski** [comment]
I think we're missing the `symfony/property-access` dependency on Resource component.

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 13 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: This is consistent with Doctrine's interfaces, let's keep it for now. :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 12 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1: for $count.

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 11 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Staying inline with Doctrine repo. For now. :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 9 above)
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1:

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 8 above)
Comment (NeverResponse) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Yup, and tested the performance, but in this case `array_filter` is slower than combining `array_search` with `unset`.

Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: :+1: let's do it this way then :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 5 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: `$interface` is a string I don't think we can do `$string instanceof ResourceInterface::class`, can we?

File: src/Sylius/Component/Resource/Exception/UnsupportedMethodException.php [hunk 14]
```diff
@@ -0,0 +1,26 @@
+<?php
+
+/*
+ * This file is part of the Lakion package.
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/Exception/UnsupportedMethodException.php]: Should be Sylius package :)

File: src/Sylius/Component/Resource/spec/Exception/UnsupportedMethodExceptionSpec.php [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+use PhpSpec\ObjectBehavior;
+
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class UnsupportedMethodExceptionSpec extends ObjectBehavior
+{
+    function let()
+    {
+        $this->beConstructedWith('methodName');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/spec/Exception/UnsupportedMethodExceptionSpec.php]: Add one spec for checking the message corretness

File: src/Sylius/Component/Resource/spec/Fixtures/SampleResourceInterface.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+/*
+ * This file is part of the Lakion package.
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/spec/Fixtures/SampleResourceInterface.php]: Sylius package

File: src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+        $book->getId()->willReturn(10);
+        $book->getName()->willReturn('Book');
+        $book->getRating()->willReturn(5);
+
+        $shirt->getId()->willReturn(5);
+        $shirt->getName()->willReturn('Shirt');
+
+        $this->add($book);
+        $this->add($shirt);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/spec/Repository/InMemoryRepositorySpec.php]: We shouldn't 'load' sample data in `let`. The examples should be complete, but when I look at https://github.com/Sylius/Sylius/pull/3617/files#diff-cba4107abbc619c9dfb57a6d36956923R154 it's confusing, because I don't know why `$book` is returned since it wasn't even added to the repository.

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+        $order = $orderBy[$property];
+
+        $sortable = array();
+        $results = array();
+
+        foreach ($resources as $key => $object) {
+            $sortable[$key] = $this->accessor->getValue($object, $property);
+        }
+
+        if ('ASC' === $order) {
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: I propose using constants for these values, wdyt?

**Comment by pjedrzejewski** [comment]
Remember about updating UPGRADE and CHANGELOG(s). :)

File: src/Sylius/Component/Resource/Repository/InMemoryRepository.php (hunk 18 above)
Comment (pamil) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Or we can extract this one function into `sortAscending(array $resource, $property)` and `sortDescending(array $resource, $property)`.

Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Repository/InMemoryRepository.php]: Hm... not sure if it's worth it if the rest of logic stays the same. That being said, that is not related to the usage of a constant here, hm? $order is coming from the user, so I think we should have constants here.

File: src/Sylius/Component/Resource/spec/Exception/UnexpectedTypeExceptionSpec.php [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace spec\Sylius\Component\Resource\Exception;
+
+use PhpSpec\ObjectBehavior;
+
+class UnexpectedTypeExceptionSpec extends ObjectBehavior
+{
+    function let()
+    {
+        $this->beConstructedWith('givenType', 'expectedType');
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Resource/spec/Exception/UnexpectedTypeExceptionSpec.php]: Maybe change `givenType` to `stringValue` to indicate it's string object and `expectedType` to `\ExpectedType` to indicate it's string class name, since it will be more intuitive.

File: src/Sylius/Bundle/ResourceBundle/Doctrine/ODM/PHPCR/DocumentRepository.php [hunk 20]
```diff
@@ -39,6 +40,28 @@ public function createPaginator(array $criteria = array(), array $sorting = arra
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function remove(ResourceInterface $resource)
+    {
+        if (null !== $this->find($resource->getId())) {
+            $this->dm->remove($resource);
+
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/ODM/PHPCR/DocumentRepository.php]: No need for the blank line here.

File: src/Sylius/Bundle/ResourceBundle/Doctrine/ODM/PHPCR/DocumentRepository.php [hunk 21]
```diff
@@ -39,6 +40,28 @@ public function createPaginator(array $criteria = array(), array $sorting = arra
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function add(ResourceInterface $resource)
+    {
+        $this->dm->persist($resource);
+
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/ODM/PHPCR/DocumentRepository.php]: Remove blank line. :)

File: src/Sylius/Bundle/ResourceBundle/Doctrine/ORM/EntityRepository.php [hunk 22]
```diff
@@ -101,6 +102,28 @@ public function findBy(array $criteria, array $sorting = array(), $limit = null,
     /**
      * {@inheritdoc}
      */
+    public function add(ResourceInterface $resource)
+    {
+        $this->_em->persist($resource);
+
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/ORM/EntityRepository.php]: I think we should remove them all, sorry for spam, I thought it is just 2 places. :)

File: src/Sylius/Component/Resource/Exception/Repository/ExistingResourceException.php [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * @author Jan Góralski <jan.goralski@lakion.com>
+ */
+class ExistingResourceException extends \Exception
+{
+    public function __construct()
+    {
+        parent::__construct('Given resource already exists in the repository.');
+    }
+}
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Resource/Exception/Repository/ExistingResourceException.php]: Perhaps this namespace should be the other way around? I mean, we could put it in ``Sylius\Component\Resource\Repository\Exception\ExistingResourceException``? Otherwise it is a bit messy to look for appropriate exception in folder with all exception then "category" of the service which this exception is thrown by. What do you think?

**Comment by pjedrzejewski** [comment]
@NeverResponse :+1: Please let me know what do you think about the exception's namespaces. I am OK with keeping generic exceptions in the ``Sylius\\Component\\Resource\\Exception`` but rest - I am not sure.

Generally I think the proper way is when every component has its own exception, even if it is something like UnexpectedTypeException, otherwise we are making kind of exceptions container from Resource component, only because other components depend on it - but that's not a valid reason.

**Comment by pjedrzejewski** [comment]
Restarted the build on PHP7, something with Selenium. :/

**Comment by pjedrzejewski** [comment]
Awesome work Jasiek! :+1: Thanks :)
